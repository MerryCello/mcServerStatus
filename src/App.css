.App {
  text-align: center;
}

body {
  padding-top: 7vh;
  background-color: rgb(30,30,30);
  background-image: url(./images/dirt_bg.png);
  background-position: center;
  background-repeat: repeat;
  background-size: cover;
  color: #fcfcfc;
  font-family: "Minecraft";
  font-size: 1.4rem;
}
h1 {
  font-size: 1.4rem;
  text-shadow: 4px 4px #3e3e3e;
  padding-left: 5%;
  padding-right: 5%;
}
table {
  margin-left: 5%;
  margin-right: 5%;
}
th {
  padding-bottom: 10px;
}
th, td {
  padding-left: 10px;
  padding-right: 10px;
  padding-bottom: 15px;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

.card-grid {
  display: grid;
  grid-template-columns: auto 1fr;
  padding: 3px;
}

div.card-grid:focus {
  background-color: black;
  outline: 3px solid white;
}

.card-icon {
  width: fit-content;
}

.card-icon>img {
  height: 6rem;
  width: auto;
  padding-right: 8px;
}

.server-details {
  display: flex;
  flex-direction: column;
  flex: 3 4;
  line-height: 1.5rem;
}

.card-srv-name-status {
  display: grid;
  grid-template-columns: auto auto;
  align-items: start;
  line-height: 2rem !important;
}

span.card-srv-name-status {
  justify-content: end;
  margin-left: 10px;
}

.servers-list {
  margin: 0 18% 0 18%;
  padding: 10px 12px 0 3px;
  height: 60vh;
  overflow-y: auto;
}

.servers-list-bg {
  background-color: rgba(0, 0, 0, 0.5);
  padding: 3px 0 3px 0;
}

::-webkit-scrollbar {
  width: 18px;
}

::-webkit-scrollbar-track {
  background-color: black;
}

::-webkit-scrollbar-thumb {
  background-color: #c0c0c0;
  border-right: 3px solid #808080;
  border-bottom: 3px solid #808080;
}

::-webkit-scrollbar-corner {
  background-color: transparent;
}

.button, .button:focus {
  color: #fcfcfc;
  text-shadow:4px 4px rgba(0, 0, 0, 0.5);
  background-image: url(./images/gray-texture.png);
  background-position: center;
  background-repeat: repeat;
  border-top: 3px solid #aaaaaa;
  border-left: 3px solid #aaaaaa;
  border-bottom: 6px solid #565656;
  border-right: 3px solid #565656;
  outline: 3px solid black;
  height: 7vh;
}

.button:disabled, .button:disabled:hover {
  border: none;
  background: linear-gradient(rgba(0, 0, 0, 0.5), rgba(0, 0, 0, 0.5)), url(./images/gray-texture.png);
  color: #9e9e9e;
  text-shadow: none;
  vertical-align: baseline;
  outline: 3px solid black;
}

.button:hover {
  outline: 3px solid white;
}

.options {
  display: grid;
  grid-template-columns: auto auto auto auto;
  margin: 3% 18% 0 18%;
  gap: 15px;
}

@media only screen and (max-width: 1000px) {
  .servers-list {
    margin: 0 5% 0 5%;
  }
  .options {
    display: grid;
    grid-template-columns: auto auto;
    margin: 3% 5% 0 5%;
  }
  body, h1 {
    font-size: 1.25rem;
  }
}

@media only screen and (max-width: 500px) {
  body {
    font-size: .75rem;
  }
  button {
    font-size: 1rem;
  }
  .card-icon>img {
    width: 15vw;
    height: auto;
    padding-right: 8px;
  }
  .card-srv-name-status {
    line-height: 1rem !important;
  }
}
