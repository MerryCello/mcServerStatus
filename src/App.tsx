import React from 'react';
import LandingPage from './pages/LandingPage';

function App() {
  return <LandingPage data-testid='LandingPage' />;
}

export default App;
