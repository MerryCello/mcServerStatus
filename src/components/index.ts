export * from './Button';
export * from './Loading';
export * from './ServerCard';
export * from './Signal';
