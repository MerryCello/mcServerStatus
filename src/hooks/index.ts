export * from "./useWindowDimensions";
export * from "./useButtonNavigate";
