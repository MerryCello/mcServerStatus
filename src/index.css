body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}

@font-face {
  font-family: "Minecraft";
  src: local("Minecraft"),
    url("./fonts/Minecraftia.ttf") format("truetype");
  font-weight: normal;
}

@font-face {
  font-family: "Minecraft-bold";
  src: local("Minecraft-bold"),
    url("./fonts/Minecrafter_3.ttf") format("truetype");
  font-weight: bold;
}
