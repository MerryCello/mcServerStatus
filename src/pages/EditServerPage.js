import React, { useEffect, useRef, useState } from "react";
import { useLocation } from "react-router-dom";
import Button from "../components/Button";
import { isNil } from "../utils";

const NAME_PLACEHOLDER = "Minecraft Server";

const EditServerPage = () => {
  const { state } = useLocation();
  const [nameState, setNameState] = useState(state?.name ?? NAME_PLACEHOLDER);
  const [addressState, setAddressState] = useState(state?.address ?? "");
  const doneRef = useRef(null);

  useEffect(() => {
    doneRef.current.disabled = addressState === "";
  }, []);

  const nameInputOnChange = ({ target: input }) => {
    setNameState(input.value);
  };
  const addressInputOnChange = ({ target: input }) => {
    setAddressState(input.value);
    if (input.value !== "") {
      doneRef.current.disabled = false;
    } else {
      doneRef.current.disabled = true;
    }
  };

  const nameInputOnBlur = () => {
    if (nameState === "") setNameState(NAME_PLACEHOLDER);
  };

  const doneOnClick = () => {
    // no state means that a new server status entry is to be added
    if (isNil(state)) {
      // TODO: create server status entry
    } else {
      // TODO: save server status entry
    }
    alert(
      " ⚠⚠⚠⚠⚠⚠⚠⚠\n👷‍♂️ Hard hat required 🛠\n👷‍♂️ Construction zone 🛠\n ⚠⚠⚠⚠⚠⚠⚠⚠"
    );
    setTimeout(() => (window.location.href = "/"), 100); // 100ms for Button sound fx to play
  };

  return (
    <div className="main-container edit-page-container">
      <h1>Edit Server Info</h1>
      <div className="edit-form-container column">
        <div className="inputs column">
          <label>Server Name</label>
          <input
            type={"text"}
            tabIndex={1}
            autoFocus
            onChange={nameInputOnChange}
            onBlur={nameInputOnBlur}
            value={nameState}
          />
          <label>Server Address</label>
          <input
            type={"text"}
            tabIndex={2}
            value={addressState}
            onChange={addressInputOnChange}
          />
        </div>
        <div className="buttons column">
          <Button onClick={doneOnClick} tabIndex={3} ref={doneRef}>
            Done
          </Button>
          <Button linkTo={"/"} tabIndex={4}>
            Cancel
          </Button>
        </div>
      </div>
    </div>
  );
};

export default EditServerPage;
