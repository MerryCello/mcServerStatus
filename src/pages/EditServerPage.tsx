import React, {
  ChangeEventHandler,
  FC,
  useEffect,
  useRef,
  useState,
} from 'react';
import {useLocation} from 'react-router-dom';
import {isNil} from 'lodash';
import Button from '../components/Button';
import {addUserServer, editUserServer} from '../firebase/controlers';
import {useButtonNavigate} from '../hooks';

export const NAME_PLACEHOLDER = 'Minecraft Server';

const EditServerPage: FC = () => {
  const {state: routeParam} = useLocation();
  const navigate = useButtonNavigate();
  const [nameState, setNameState] = useState(
    routeParam?.name ?? NAME_PLACEHOLDER,
  );
  const [addressState, setAddressState] = useState(routeParam?.address ?? '');
  const doneRef = useRef<{disabled: boolean | null}>(null);

  useEffect(() => {
    if (doneRef.current) {
      doneRef.current.disabled = addressState === '';
    }
  }, []);

  const nameInputOnChange: ChangeEventHandler<HTMLInputElement> = ({
    target: input,
  }) => {
    setNameState(input.value);
  };
  const addressInputOnChange: ChangeEventHandler<HTMLInputElement> = ({
    target: input,
  }) => {
    setAddressState(input.value);
    if (doneRef.current) {
      if (input.value !== '') {
        doneRef.current.disabled = false;
      } else {
        doneRef.current.disabled = true;
      }
    }
  };

  const nameInputOnBlur = () => {
    if (nameState === '') setNameState(NAME_PLACEHOLDER);
  };

  const doneOnClick = () => {
    let resolution = Promise.resolve();
    const aNewServerIsAdded = isNil(routeParam);
    const anInputFieldHasChanged =
      routeParam?.address !== addressState || routeParam?.name !== nameState;

    if (aNewServerIsAdded) {
      resolution = addUserServer({
        address: addressState,
        name: nameState,
      });
    } else if (anInputFieldHasChanged) {
      resolution = editUserServer({
        id: routeParam?.id,
        address: addressState,
        name: nameState,
      });
    }
    resolution.then(() => navigate('/mcServerStatus'));
  };

  return (
    <div className='main-container edit-page-container'>
      <h1 data-testid='page-title'>Edit Server Info</h1>
      <div className='edit-form-container column'>
        <div className='inputs column'>
          <label data-testid='name-label'>Server Name</label>
          <input
            data-testid='name-input'
            type={'text'}
            tabIndex={1}
            autoFocus
            onChange={nameInputOnChange}
            onBlur={nameInputOnBlur}
            value={nameState}
          />
          <label data-testid='address-label'>Server Address</label>
          <input
            data-testid='address-input'
            type={'text'}
            tabIndex={2}
            value={addressState}
            onChange={addressInputOnChange}
          />
        </div>
        <div className='buttons column'>
          <Button
            data-testid='done-button'
            onClick={doneOnClick}
            tabIndex={3}
            ref={doneRef}>
            Done
          </Button>
          <Button
            data-testid='cancel-button'
            linkTo={'/mcServerStatus'}
            tabIndex={4}>
            Cancel
          </Button>
        </div>
      </div>
    </div>
  );
};

export default EditServerPage;
