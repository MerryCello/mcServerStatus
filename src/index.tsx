import React from 'react';
import ReactDOM from 'react-dom/client';
import {createBrowserRouter, RouterProvider, Navigate} from 'react-router-dom';
import './index.css';
import App from './App';
import reportWebVitals from './reportWebVitals';
import NotFound404 from './pages/NotFound404';
import EditServerPage from './pages/EditServerPage';
import DeleteServerPage from './pages/DeleteServerPage';

const router = createBrowserRouter([
  {
    path: '*',
    element: <Navigate to='/mcServerStatus' replace />,
    errorElement: <NotFound404 />,
  },
  {
    path: "/mcServerStatus",
    element: <App />,
    errorElement: <NotFound404 />,
  },
  {
    path: '/mcServerStatus/edit',
    element: <EditServerPage />,
    errorElement: <NotFound404 />,
  },
  {
    path: '/edit',
    element: <Navigate to='/mcServerStatus/edit' replace />,
    errorElement: <NotFound404 />,
  },
  {
    path: '/mcServerStatus/add',
    element: <EditServerPage />,
    errorElement: <NotFound404 />,
  },
  {
    path: '/add',
    element: <Navigate to='/mcServerStatus/add' replace />,
    errorElement: <NotFound404 />,
  },
  {
    path: '/mcServerStatus/delete',
    element: <DeleteServerPage />,
    errorElement: <NotFound404 />,
  },
  {
    path: '/delete',
    element: <Navigate to='/mcServerStatus/delete' replace />,
    errorElement: <NotFound404 />,
  },
]);

ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
  <React.StrictMode>
    <RouterProvider router={router} />
  </React.StrictMode>,
);

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals();
