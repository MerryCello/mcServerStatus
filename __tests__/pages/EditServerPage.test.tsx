/** @jest-environment jsdom */
import React, {useState} from 'react';
import {fireEvent, render} from '@testing-library/react';
import {
  /*addUserServer, */ editUserServer,
} from '../../src/firebase/controlers';
import {useButtonNavigate} from '../../src/hooks';
import EditServerPage /*NAME_PLACEHOLDER*/ from '../../src/pages/EditServerPage';
import {useLocation} from 'react-router-dom';
import {noop} from 'lodash';

jest.mock('../../src/firebase/controlers');
jest.mock('../../src/hooks');
// const buttonRef = {current: {disabled: null}};
jest.mock('../../src/components/Button', () => (props) => (
  // @ts-expect-error
  <mc-button {...props} />
));
jest.spyOn(console, 'error');

describe('EditServerPage', () => {
  const serverToEdit = {
    name: 'aServer',
    address: '123.123.123.123',
    id: 'test-id',
  };
  const editedServer = {
    name: 'a Server',
    address: '321.312.312.312',
    id: 'test-id',
  };

  beforeEach(jest.clearAllMocks);

  it('should render to edit an existing server', () => {
    const mockSetNameState = jest.fn();
    const mockSetAddressState = jest.fn();
    (useLocation as jest.Mock).mockReturnValueOnce({state: serverToEdit});
    (useState as jest.Mock)
      .mockReturnValueOnce([serverToEdit.name, mockSetNameState])
      .mockReturnValueOnce([serverToEdit.address, mockSetAddressState]);

    const {container, getByTestId} = render(<EditServerPage />);

    expect(getByTestId('page-title')).toBeDefined();

    // name
    expect(getByTestId('name-label')).toBeDefined();
    const nameInput = getByTestId('name-input');
    expect(nameInput).toHaveAttribute('value', serverToEdit.name);
    fireEvent.change(nameInput, {target: {value: editedServer.name}});
    // fireEvent.blur(nameInput, {target: {value: 'ABC'}}); // TODO: for new server test
    expect(mockSetNameState).toHaveBeenCalledWith(editedServer.name);

    // address
    expect(getByTestId('address-label')).toBeDefined();
    const addressInput = getByTestId('address-input');
    expect(addressInput).toHaveAttribute('value', serverToEdit.address);
    fireEvent.change(addressInput, {target: {value: editedServer.address}});
    expect(mockSetAddressState).toHaveBeenCalledWith(editedServer.address);

    expect(container).toMatchSnapshot();
  });

  it('should call editUserServer and navigate back when the done button is clicked, and name and address have changed', () => {
    const mockNavigate = jest.fn();
    const mockSetNameState = jest.fn();
    const mockSetAddressState = jest.fn();

    (useLocation as jest.Mock).mockReturnValueOnce({state: serverToEdit});
    (useButtonNavigate as jest.Mock).mockReturnValueOnce(mockNavigate);
    (useState as jest.Mock)
      .mockReturnValueOnce([editedServer.name, mockSetNameState])
      .mockReturnValueOnce([editedServer.address, mockSetAddressState]);
    (editUserServer as jest.Mock).mockImplementationOnce(() => ({
      then: (fn) => fn(),
    }));

    const {getByTestId} = render(<EditServerPage />);

    const doneButton = getByTestId('done-button');
    fireEvent.click(doneButton);
    expect(editUserServer).toHaveBeenCalledWith(
      editedServer,
      undefined,
      undefined,
    );
    expect(mockNavigate).toBeCalledWith(-1);
  });

  it('should navigate back with no changes made when done button is clicked', () => {
    const mockNavigate = jest.fn();
    const mockThenCall = () => ({
      then: (fn) => fn(),
    });
    // @ts-ignore
    jest.spyOn(Promise, 'resolve').mockImplementation(mockThenCall);
    (useLocation as jest.Mock).mockReturnValueOnce({state: serverToEdit});
    (useButtonNavigate as jest.Mock).mockReturnValueOnce(mockNavigate);
    (useState as jest.Mock)
      .mockReturnValueOnce([serverToEdit.name, noop])
      .mockReturnValueOnce([serverToEdit.address, noop]);
    (editUserServer as jest.Mock).mockImplementationOnce(mockThenCall);

    const {getByTestId} = render(<EditServerPage />);

    const doneButton = getByTestId('done-button');
    fireEvent.click(doneButton);
    expect(editUserServer).not.toHaveBeenCalled();
    expect(mockNavigate).toHaveBeenCalledWith(-1);
  });

  it('should render to add a new server', () => {
    const mockSetNameState = jest.fn();
    const mockSetAddressState = jest.fn();
    (useLocation as jest.Mock).mockReturnValueOnce({state: {}});
    (useState as jest.Mock)
      .mockReturnValueOnce(['', mockSetNameState])
      .mockReturnValueOnce(['', mockSetAddressState]);

    const {container, getByTestId} = render(<EditServerPage />);

    expect(getByTestId('page-title')).toBeDefined();

    // name
    expect(getByTestId('name-label')).toBeDefined();
    const nameInput = getByTestId('name-input');
    expect(nameInput).toHaveAttribute('value', '');
    fireEvent.change(nameInput, {target: {value: 'New Server'}});
    expect(mockSetNameState).toHaveBeenCalledWith('New Server');

    // address
    expect(getByTestId('address-label')).toBeDefined();
    const addressInput = getByTestId('address-input');
    expect(addressInput).toHaveAttribute('value', '');
    fireEvent.change(addressInput, {target: {value: '123.123.123.123'}});
    expect(mockSetAddressState).toHaveBeenCalledWith('123.123.123.123');

    expect(container).toMatchSnapshot();
  });

  it('should send the sharedListId when editing a server for a shared list', () => {
    const mockNavigate = jest.fn();
    const mockSetNameState = jest.fn();
    const mockSetAddressState = jest.fn();
    const sharedListId = 'shared-list-id';

    (useLocation as jest.Mock).mockReturnValueOnce({
      state: {...serverToEdit, listId: sharedListId},
    });
    (useButtonNavigate as jest.Mock).mockReturnValueOnce(mockNavigate);
    (useState as jest.Mock)
      .mockReturnValueOnce([editedServer.name, mockSetNameState])
      .mockReturnValueOnce([editedServer.address, mockSetAddressState]);
    (editUserServer as jest.Mock).mockImplementationOnce(() => ({
      then: (fn) => fn(),
    }));

    const {getByTestId} = render(<EditServerPage />);

    const doneButton = getByTestId('done-button');
    fireEvent.click(doneButton);
    expect(editUserServer).toHaveBeenCalledWith(
      editedServer,
      undefined,
      sharedListId,
    );
    expect(mockNavigate).toBeCalledWith(-1);
  });

  it('should place default server name when input name is blank', () => {
    const mockSetNameState = jest.fn();
    const mockSetAddressState = jest.fn();

    (useLocation as jest.Mock).mockReturnValueOnce({state: {}});
    (useState as jest.Mock)
      .mockReturnValueOnce(['', mockSetNameState])
      .mockReturnValueOnce(['', mockSetAddressState]);

    const {getByTestId} = render(<EditServerPage />);

    const nameInput = getByTestId('name-input');
    fireEvent.blur(nameInput, {target: {value: ''}});
    expect(mockSetNameState).toHaveBeenCalledWith('Minecraft Server');
  });

  it('should navigate back when cancel is pressed', () => {
    const mockNavigate = jest.fn();
    (useButtonNavigate as jest.Mock).mockReturnValueOnce(mockNavigate);
    (useLocation as jest.Mock).mockReturnValueOnce({state: serverToEdit});

    const {getByTestId} = render(<EditServerPage />);

    const cancelButton = getByTestId('cancel-button');
    fireEvent.click(cancelButton);
    expect(mockNavigate).toHaveBeenCalledWith(-1);
  });
});
