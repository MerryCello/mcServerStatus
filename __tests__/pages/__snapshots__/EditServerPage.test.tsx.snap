// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditServerPage should render to edit an existing server 1`] = `
<div>
  <div
    class="main-container edit-page-container"
  >
    <h1
      data-testid="page-title"
    >
      Edit Server Info
    </h1>
    <div
      class="edit-form-container column"
    >
      <div
        class="inputs column"
      >
        <label
          data-testid="name-label"
        >
          Server Name
        </label>
        <input
          data-testid="name-input"
          tabindex="1"
          type="text"
          value="aServer"
        />
        <label
          data-testid="address-label"
        >
          Server Address
        </label>
        <input
          data-testid="address-input"
          tabindex="2"
          type="text"
          value="123.123.123.123"
        />
      </div>
      <div
        class="buttons column"
      >
        <mc-button
          data-testid="done-button"
          tabindex="3"
        >
          Done
        </mc-button>
        <mc-button
          data-testid="cancel-button"
          linkto="/mcServerStatus"
          tabindex="4"
        >
          Cancel
        </mc-button>
      </div>
    </div>
  </div>
</div>
`;
